js/src/components/tree-select-control/index.js

```diff
@@ -1,56 +1,217 @@
+			handleParentChange( checked, option );
+		} else {
+			handleSingleChange( checked, option );
+		}
+	};
+
+	/**
+	 * Handles a change of a child element.
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleSingleChange = ( checked, option ) => {
+		const newValue = checked
+			? [ ...value, option.id ]
+			: value.filter( ( el ) => el !== option.id );
+
+		onChange( newValue );
+	};
+
+	/**
+	 * Handles a change of a Parent element.
+	 *
+	 * @param {boolean} checked Indicates if the item should be checked
+	 * @param {Option} option The option to change
+	 */
+	const handleParentChange = ( checked, option ) => {
+		const newValue = [ ...value ];
+
+		function loadChildren( parent ) {
+			if ( ! parent.children ) {
+				return;
+			}
+
+			parent.children.forEach( ( child ) => {
+				if ( child.children?.length ) {
+					loadChildren( child );
+					return;
+				}
+
+				const childIdPosition = newValue.indexOf( child.id );
+
+				if ( ! checked && childIdPosition >= 0 ) {
+					newValue.splice( childIdPosition, 1 );
+				}
+
+				if ( checked && childIdPosition < 0 ) {
+					newValue.push( child.id );
+				}
+			} );
+		}
+
+		loadChildren( option );
+		onChange( newValue );
+	};
+
+	/**
+	 * Handles a change of a Tag element. We map them to Value format.
+	 *
+	 * @param {Array} tags List of current tags
+	 */
+	const handleTagsChange = ( tags ) => {
+		onChange( [ ...tags.map( ( el ) => el.id ) ] );
+	};
 
 	return (
 		<div
+			{ ...focusOutside }
 			className={ classnames(
 				'woocommerce-tree-select-control',
 				className
 			) }
 		>
+			{ !! label && (
+				<label
+					htmlFor={ `woocommerce-tree-select-control-${ instanceId }__control-input` }
+					className="woocommerce-tree-select-control__label"
+				>
+					{ label }
+				</label>
+			) }
+
 			<Control
-				disabled={ disabled }
-				hasTags
+				disabled={ disabled || ! options.length }
+				tags={ getTags() }
 				isExpanded={ isExpanded }
-				// onSearch={ this.search }
-				// selected={ this.getSelected() }
-				onChange={ onChange }
-				setExpanded={ setIsExpanded }
-				// updateSearchOptions={ this.updateSearchOptions }
-				// decrementSelectedIndex={ this.decrementSelectedIndex }
-				// incrementSelectedIndex={ this.incrementSelectedIndex }
+				onFocus={ () => {
+					setIsExpanded( true );
+				} }
+				instanceId={ instanceId }
+				placeholder={ placeholder }
+				label={ label }
+				onTagsChange={ handleTagsChange }
 			/>
-			{ /* { isExpanded && (
+			{ isExpanded && (
 				<List
-					node={ this.node }
-					onSelect={ this.selectOption }
-					onSearch={ this.search }
-					options={ this.getOptions() }
-					decrementSelectedIndex={ this.decrementSelectedIndex }
-					incrementSelectedIndex={ this.incrementSelectedIndex }
-					setExpanded={ this.setExpanded }
+					options={ options }
+					value={ value }
+					onChange={ handleListChange }
```

Comment:
❔ 💅 Throwing out an idea here, not sure if it is feasible, something that we can consider later. This is related to the `onChange` and the internals of the `List` and `Options` components.

In `List`, for `value` props, we pass in an array of `id`, and for `onChange`, _it gives us `checked` and `option` node_, and in the `handleListChange` event handler, _we process them with `handleParentChange` and `handleSingleChange`_, resulting in `newValue` and pass it to consumer via `onChange`.

I'm thinking, instead of `checked` and `option` parameters, can the `onChange` always give us `value` array? Instead of us processing the values with `handleParentChange` and `handleSingleChange`, perhaps we can do it another way using components approach, like example below?

```js
	return options.map( ( option ) => {
		return option.children?.length ? (
			<ParentNode option={ option } onChange={ handleParentNodeChange } />
		) : (
			<Node option={ option } onChange={ handleNodeChange } />
		);
	} );
```

So the idea above is to have different components for parent node and leaf node, and with that, we can have different handlers for the parent node and leaf node. In a way, I think that may be clearer. 

Also, note that in later iterations, the parent node would need to have the expand/collapse functionality, so having the different ParentNode and leaf node components can be helpful in UI and CSS later too.

One possible drawback about having all `onChange` to pass `value` is that the performance may suffer, especially when we have a lot of levels with many parent nodes, each level would need to have their `handleParentNodeChange` to process and pass the new value to their parent.

Just something to think about. 😅

Reply:
Side note: I looked into Ant Design TreeSelect component and I noticed that their tree nodes are actually rendered in sequential list of divs, instead of nested divs. I wonder if that is also a way to address performance. That may be an angle that we can look at.

![image](https://user-images.githubusercontent.com/417342/156030049-96160ae0-fdf8-4f22-956f-a3fe1a335ceb.png)

Reply:
Those are interesting points @ecgan 

One thing we can do is to use Divide Milestone 2.

**Milestone 2.A** 

expand/collapse nodes, search, "Select all options"  and performance. 

The performance will include a stress test and we can experiment with these options you provide to compare. I think we need to decide about the structure as soon as possible in the project, to avoid heavy refactoring.

**Milestone 2.B**

Accessibility and final style tweaks


WDYT?

Reply:
@puntope , I think that sounds good. 🙂 

Just to be sure, do you mean to have one PR for Milestone 2.A, and another PR for Milestone 2.B? To me, it's fine to have more PRs. The number of PRs can be more than the number of milestones. We can have one PR for expand/collapse nodes, one PR for search, etc.